warp/yul/ScopeFlattener.py

```diff
@@ -47,13 +47,78 @@ def visit_block(self, node: ast.Block, inline: Optional[bool] = None):
         return ast.Block((self.visit(block_call),))
 
     def visit_function_definition(self, node: ast.FunctionDefinition):
+        # We do not want to flatten an if-block if it is the only statement in
+        # the function body. Skip visit_if in such cases and instead visit all
+        # the branches of the if-block.
+        if len(node.body.statements) == 1 and isinstance(
+            node.body.statements[0], ast.If
+        ):
+            if_statement = node.body.statements[0]
+
+            return ast.FunctionDefinition(
+                name=node.name,
+                parameters=self.visit_list(node.parameters),
+                return_variables=self.visit_list(node.return_variables),
+                body=ast.Block(
+                    (
+                        ast.If(
+                            condition=self.visit(if_statement.condition),
+                            body=self.visit(if_statement.body),
+                            else_body=self.visit(if_statement.else_body)
+                            if if_statement.else_body
+                            else None,
+                        ),
+                    )
+                ),
+            )
+
         return ast.FunctionDefinition(
             name=node.name,
             parameters=self.visit_list(node.parameters),
             return_variables=self.visit_list(node.return_variables),
             body=self.visit(node.body, inline=True),
         )
 
+    def visit_if(self, node: ast.If):
+        free_vars = set()
+        mod_vars = set()
+
+        free_vars.update(node.body.scope.free_variables)
+        free_vars.update(node.body.scope.modified_variables)
+        mod_vars.update(node.body.scope.modified_variables)
+
+        free_vars.update(ast.Block((node.condition,)).scope.free_variables)
+
+        if node.else_body:
+            free_vars.update(node.else_body.scope.free_variables)
+            free_vars.update(node.else_body.scope.modified_variables)
+            mod_vars.update(node.else_body.scope.modified_variables)
```

Comment:
A curious thought: you can get all free and modified variables in a more terse way by simply wrapping the whole `node` into a `Block`. Then you call `.scope` and here you are

Reply:
Yea that seems like a better way; I'll add those changes